File: solutions/ruby/simple-cipher/1/simple_cipher.rb
# frozen_string_literal: true

# class Cipher
class Cipher
  attr_reader :key

  def initialize(key = nil)
    @key = key || Array.new(100) { rand(97..122) }.map(&:chr).join
    raise ArgumentError unless @key =~ /[a-z]+/
  end

  def encode(plaintext) = shift(plaintext) { _1 + _2 }
  def decode(ciphertext) = shift(ciphertext) { _1 - _2 }

  private

  def shift(text, &block) =
    text.each_char.zip(@key.chars)
        .map { (block.call(_1.ord - 97, _2.ord - 97) % 26 + 97).chr }
        .join
end

File: solutions/clojure/simple-cipher/1/src/simple_cipher.clj
(ns simple-cipher)

(def ^:private a-code (int \a))
(def ^:private alphabet-size 26)

(defn rand-key [] (apply str (repeatedly 100 #(char (+ a-code (rand-int alphabet-size))))))

(defn- transform [op key text]
  (apply str
         (map #(char (+ a-code
                        (mod (op (- (int %1) a-code)
                                 (- (int %2) a-code))
                             alphabet-size)))
              text
              (cycle key))))

(defn encode [key text] (transform + key text))
(defn decode [key text] (transform - key text))
